### PR TITLE
GEOS-9239: optional flag(s) to remove root layer from WMS capabilitie…

### PR DESCRIPTION
**Pull request by mbarto**
…s document when possible

## Description

To increase compatibility with WMS clients, we introduce some options to remove the Root Layer element from WMS capabilities documents when this is possible (there is a single group or layer as the only top level element).
This can be set at the service, layer (or group), or request level.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)

The following are required only for core and extension modules, while they are welcomed, but not required, for community modules:
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [x] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

**Comment by aaime** [comment]
Not sure why Travis build is not showing up here, but it's in progress, and failed on the QA bits, link here:
https://travis-ci.org/geoserver/geoserver/builds/543771385

**Comment by mbarto** [comment]
Missing comment, should be ok now

File: doc/en/user/source/security/tutorials/ldap/index.rst [hunk 1]
```diff
@@ -165,7 +165,7 @@ from the LDAP repository and allow access rights to be assigned to those roles.
 
 #. Click the ``LDAP`` option under the  ``New Role Service`` section.
 
-   .. figure:: images/ldap15.jpg
+   .. figure:: images/ldap_15.jpg
```
Comment (jodygarnett) [inline doc/en/user/source/security/tutorials/ldap/index.rst]: This appears to be an unrelated change?

File: doc/en/user/source/security/webadmin/auth.rst [hunk 2]
```diff
@@ -226,7 +226,7 @@ The following illustration shows the configuration options for the LDAP authenti
      - Search pattern for locating the LDAP groups a user belongs to. This may contain two placeholder values:
        ``{0}``, the full DN of the user, for example ``uid=bob,ou=people,dc=acme,dc=com``
        ``{1}``, the ``uid`` portion of the full DN, for example ``bob``.
-       Only applicable when the *Use LDAP groups for authorization( parameter is **checked**.
+       Only applicable when the *Use LDAP groups for authorization* parameter is **checked**.
```
Comment (jodygarnett) [inline doc/en/user/source/security/webadmin/auth.rst]: Looks to be a documentation fix.

**Comment by mbarto** [comment]
@jodygarnett I removed the unrelated docs changes, I will make a separate PR for thise

**Comment by aaime** [comment]
Trying to build the docs for this I got a lot of warnings and the doc build eventually failed, but the warnings are about a lot of files that @mbarto did not touch. @jodygarnett you made some changes in the doc build lately didn't you? Would those explain the build failure?

Both "ant user" and "mvn install" fail, here is the output for the second:

```
sphinx:
     [echo] Running sphinx-build -D release=2.16-SNAPSHOT -W -b html -d "/home/aaime/devel/git-gs/doc/en/target/docguide/doctrees" . "/home/aaime/devel/git-gs/doc/en/target/docguide/html"

user:

sphinx:
     [echo] Running sphinx-build -D release=2.16-SNAPSHOT -W -b html -d "/home/aaime/devel/git-gs/doc/en/target/user/doctrees" . "/home/aaime/devel/git-gs/doc/en/target/user/html"
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/googleearth/quickstart.rst:4: WARNING: duplicate label google_earth_quickstart, other instance in /home/aaime/devel/git-gs/doc/en/user/source/services/wms/googleearth/quickstart.rst
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/googleearth/quickstart.rst:: WARNING: image file not readable: googleearth/../webadmin/images/preview_list.png
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/data/webadmin/layergroups.rst:145: WARNING: image file not readable: data/webadmin/img/data_layers_root_in_capabilities.png
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/data/webadmin/layers.rst:233: WARNING: image file not readable: data/webadmin/img/data_layers_root_in_capabilities.png
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/googleearth/tutorials/superoverlaysgwc.rst:4: WARNING: duplicate label ge-tutorial-superoverlays-gwc, other instance in /home/aaime/devel/git-gs/doc/en/user/source/services/wms/googleearth/tutorials/superoverlaysgwc.rst
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/googleearth/tutorials/superoverlaysgwc.rst:: WARNING: image file not readable: googleearth/tutorials/../../webadmin/tilecache/img/demopage.png
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/advanced/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/community/dxf/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/community/libjpeg-turbo/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/community/netcdf-out/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/community/netcdf/netcdf.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/community/printing/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/community/python/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/data/raster/imagemosaic.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/data/vector/vpf.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/datadirectory/data-dir-creating.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/datadirectory/data-dir-setting.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/datadirectory/data-dir-structure.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/extensions/css/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/extensions/csw/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/extensions/geosearch.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/extensions/monitoring/db.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/extensions/monitoring/upgrade.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/extensions/styler.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/extensions/wps/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/gettingstarted/styling-quickstart/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/googleearth/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/installation/linux/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/installation/osx/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/installation/windows/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/rest/examples/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/services/virtual-services.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/services/wms/get_legend_graphic/legendgraphic.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/styling/sld-cookbook/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/styling/sld-extensions/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/styling/sld-introduction.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/styling/sld-reference/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/styling/sld-tipstricks/index.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/styling/sld-working.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/tutorials/image_mosaic_plugin/imagemosaic.rst: WARNING: document isn't included in any toctree
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/googleearth/quickstart.rst:11: WARNING: undefined label: webadmin_basics (if the link has no caption the label must precede a section header)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  11.373 s
[INFO] Finished at: 2019-06-12T17:18:46+02:00
[INFO] ------------------------------------------------------------------------
```

**Comment by aaime** [comment]
@mbarto there is at least one issue what seems relevant to you changes though, a missing image:

```
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/data/webadmin/layergroups.rst:145: WARNING: image file not readable: data/webadmin/img/data_layers_root_in_capabilities.png
     [exec] /home/aaime/devel/git-gs/doc/en/user/source/data/webadmin/layers.rst:233: WARNING: image file not readable: data/webadmin/img/data_layers_root_in_capabilities.png
```

**Comment by mbarto** [comment]
@aaime you are right, I forgot to commit that one, committed now

File: src/wms/src/main/java/org/geoserver/wms/WMS.java [hunk 3]
```diff
@@ -218,6 +218,12 @@
     /** DateLine Wrapping Heuristic key */
     public static String DATELINE_WRAPPING_HEURISTIC_KEY = "disableDatelineWrappingHeuristic";
 
+    /** Capabilities will be produced without a root Layer element, when not needed * */
+    public static Boolean NO_ROOT_LAYER_IN_CAPABILITIES = false;
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/WMS.java]: "NO_ROOT_LAYER_IN_CAPS_DEFAULT" maybe (was not clear to me reading the code that it was the default value, tried to make the name a bit shorter to compensate)

File: src/wms/src/test/java/org/geoserver/wms/capabilities/CapabilitiesXmlReaderTest.java [hunk 4]
```diff
@@ -39,5 +41,58 @@ public void testParseXmlGetCapabilities() throws Exception {
...
+    public void testParseXmlGetCapabilitiesNoRootLayerTrue() throws Exception {
+        CapabilitiesXmlReader reader =
+                new CapabilitiesXmlReader(EntityResolverProvider.RESOLVE_DISABLED_PROVIDER);
+
+        String plainRequest =
+                "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" //
+                        + "<ogc:GetCapabilities xmlns:ogc=\"http://www.opengis.net/ows\" " //
+                        + "         xmlns:gml=\"http://www.opengis.net/gml\" " //
+                        + "         version=\"1.2.0\" updateSequence=\"1\" " //
+                        + "         noRootLayer=\"true\" " //
```
Comment (aaime) [inline src/wms/src/test/java/org/geoserver/wms/capabilities/CapabilitiesXmlReaderTest.java]: Seeing here and everywhere else, negations everywhere... would it had been better to have a rootLayer attribute/flag everywhere defaulting to true instead? But I understand it would make for a likely wide change in the PR....

File: src/wms/src/main/java/org/geoserver/wms/capabilities/Capabilities_1_3_0_Transformer.java [hunk 5]
```diff
@@ -121,19 +122,43 @@
...
+     * @param extCapsProviders collection of providers of extended capabilities content
+     * @param noRootLayer whether to skip root Layer element if there is a already single top Layer
+     *     element
+     */
+    public Capabilities_1_3_0_Transformer(
+            WMS wms,
+            String schemaBaseUrl,
+            Collection<GetMapOutputFormat> getMapFormats,
+            Collection<ExtendedCapabilitiesProvider> extCapsProviders,
+            Boolean noRootLayer) {
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/capabilities/Capabilities_1_3_0_Transformer.java]: This new optional parameter caused many changes in the code, with nulls added at the end of various constructor calls... wouldn't it be better to make it a setter and invoke it only when needed instead, leaving the constructor as it was?

File: src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/CapabilitiesTest.java [hunk 6]
```diff
@@ -795,6 +796,223 @@ public void testNonAdvertisedNonQueriableWithinGroup() throws Exception {
...
 
+    @Test
+    public void testRootLayerRemovedWMSService() throws Exception {
+        // make layers non advertised
+        WMSInfo info = getWMS().getServiceInfo();
+        info.getMetadata().put(WMS.NO_ROOT_LAYER_IN_CAPABILITIES_KEY, true);
+        getGeoServer().save(info);
+        Catalog catalog = getCatalog();
+        setAdvertised(catalog, MockData.LAKES, false);
+        setQueryable(catalog, MockData.LAKES, false);
```
Comment (aaime) [inline src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/CapabilitiesTest.java]: I don't understand why the queriable flag has been switched off here? Does it affect layer visibility in the capabilities document?
Also, is making the layers not advertised changing the outcome? They are part of a "named" group, not a "single" group

File: src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/CapabilitiesTest.java [hunk 7]
```diff
@@ -795,6 +796,223 @@ public void testNonAdvertisedNonQueriableWithinGroup() throws Exception {
...
+        setAdvertised(catalog, MockData.NAMED_PLACES, false);
+        setQueryable(catalog, MockData.NAMED_PLACES, false);
+        LayerGroupInfo group = null;
+        try {
+            group = createLakesPlacesLayerGroup(catalog, LayerGroupInfo.Mode.NAMED, null);
+            group.getMetadata().put(PublishedInfo.ROOT_IN_CAPABILITIES, true);
+            Document dom =
+                    dom(
+                            get(
+                                    "lakes_and_places/wms?request=GetCapabilities&version=1.1.0&noRootLayer=false"),
```
Comment (aaime) [inline src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/CapabilitiesTest.java]: ``noRootLayer=false`` is hard to read imho, it means that the root layer is there, double negation.
``rootLayer=true``, ``rootLayer=false`` seem to read better to me.

File: src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/CapabilitiesTest.java [hunk 8]
```diff
@@ -795,6 +796,223 @@ public void testNonAdvertisedNonQueriableWithinGroup() throws Exception {
...
+        setQueryable(catalog, MockData.LAKES, false);
+        LayerInfo layer = catalog.getLayerByName(MockData.LAKES.getLocalPart());
+
+        try {
+            layer.getMetadata().put(PublishedInfo.ROOT_IN_CAPABILITIES, true);
+            catalog.save(layer);
+            Document dom = dom(get("cite/Lakes/wms?request=GetCapabilities&version=1.1.0"), true);
+            // print(dom);
+
+            assertXpathEvaluatesTo("Lakes", "/WMT_MS_Capabilities/Capability/Layer/Name", dom);
```
Comment (aaime) [inline src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/CapabilitiesTest.java]: But if the layer is not advertised, it should not show up in the capabilities, the document should be empty....

File: src/wms/src/test/java/org/geoserver/wms/wms_1_3/CapabilitiesIntegrationTest.java [hunk 9]
```diff
@@ -945,6 +946,181 @@ public void testLayerGroupsPerWorkspace() throws Exception {
...
+            if (group != null) {
+                catalog.remove(group);
+            }
+            info.getMetadata().remove(WMS.NO_ROOT_LAYER_IN_CAPABILITIES_KEY);
+        }
+    }
+
+    @Test
+    public void testRootLayerRemovedGroupConfig() throws Exception {
+        // make layers non advertised
```
Comment (aaime) [inline src/wms/src/test/java/org/geoserver/wms/wms_1_3/CapabilitiesIntegrationTest.java]: Actually it does not make them non advertised and/or not queriable... maybe this version of the tests (1.3 vs 1.1) got rid of the unecessary bits? These tests are much easier to read!

Comment (aaime) [reply src/wms/src/test/java/org/geoserver/wms/wms_1_3/CapabilitiesIntegrationTest.java]: Also, maybe it's interesting to see a test where a bunch of layers but one got marked as non advertised (hence not visible in the caps) and verify that the lone remaining layer got moved to the top

File: doc/en/user/source/data/webadmin/layers.rst [hunk 10]
```diff
@@ -218,6 +218,20 @@ Cache parameters that apply to the HTTP response from client requests.
...
 
+Root Layer in Capabilities
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When a layer is the only top level element in the Capabilities document, it is possible to remove the Root Layer that contains it, and return
+a hierarchy where the layer is the root instead.
+
+To enable this functionality, choose the **No** option from the Root Layer in Capabilities section.
+By default this behaviour is inherited from the global WMS service settings (**WMS Global Settings** option).
+Finally, it is possible to override the service settings and force a **Yes** to always include the GeoServer root element.
```
[the remark this answers is not in the document]
Comment (jodygarnett) [reply doc/en/user/source/data/webadmin/layers.rst]: This may be easier to understand with a check box (on by default) to enable root layer, and when checked the fields to fill in description are enabled. 

"to enable select *no*" is the problem here, expressed as a double negative.

Can we express it as a positive and show what the effect of enable / disable will be?

File: src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilitiesXmlReader.java [hunk 11]
```diff
@@ -124,6 +124,8 @@ public void startElement(
                         request.setVersion(version);
                     } else if (atts.getLocalName(i).equals("updateSequence")) {
                         request.setUpdateSequence(atts.getValue(i));
+                    } else if (atts.getLocalName(i).equals("noRootLayer")) {
```
[the remark this answers is not in the document]
Comment (jodygarnett) [reply src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilitiesXmlReader.java]: Is this worth it for what is a vendor option override? What is the goal here force show or force hide the root layer on a client by client basis?

File: doc/en/user/source/data/webadmin/layers.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (mbarto) [reply doc/en/user/source/data/webadmin/layers.rst]: @jodygarnett I agree that inverting the logic is probably better, I will try to do that (also the checkbox is a good idea).
@aaime the 3 levels of configuration mimick what we already do for interpolation settings, where we can have a WMS service default, a layer layer override and a request level override.
They work in this order: WMS config is the default, layer (or group) config overrides the default if set to Yes or No, request parameter, if present,  wins over both.
I will try to improve documentation too.

File: src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilitiesXmlReader.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (mbarto) [reply src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilitiesXmlReader.java]: Support for POST is not needed by customer, if you prefer I can remove it

File: src/main/src/main/java/org/geoserver/catalog/PublishedInfo.java [hunk 12]
```diff
@@ -14,6 +14,9 @@
  */
 public interface PublishedInfo extends CatalogInfo {
 
+    /** Use the entity as capabilities root element if it's the only top level element */
+    public static final String ROOT_IN_CAPABILITIES = "rootInCapabilities";
```
[the remark this answers is not in the document]
Comment (mbarto) [reply src/main/src/main/java/org/geoserver/catalog/PublishedInfo.java]: This is actually a metadata setting for layer and layergroups, so I put it in the base info class of both (PublishedInfo)

[the remark this answers is not in the document]
Comment (aaime) [reply src/main/src/main/java/org/geoserver/catalog/PublishedInfo.java]: Yeah, I realized later that it was used in this place too, forgot to remove the comment.

File: src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilitiesXmlReader.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (aaime) [reply src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilitiesXmlReader.java]: It would not be the first option that is available in KVP only, however users have complained in the past about this asymmetry (e.g. viewParams in WFS GET but not WFS POST, however for example I don't believe you can pass viewParams in WMS GetMap POST). I'm open to both approaches (documenting, or removing).

File: src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/CapabilitiesTest.java (hunk 8 above)
Comment (mbarto) [reply src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/CapabilitiesTest.java]: @aaime About this one, I think there is a bug in WMS 1.1.1 capabilities, it looks like advertised flag is not considered at all (while it is in 1.3.0). I would eventually open a separate JIRA for this. In the meantime I am going to cleanup my tests a bit (advertised / queryable was a copy/paste side effect, not needed for the tests themselves).

File: src/wms/src/test/java/org/geoserver/wms/wms_1_3/CapabilitiesIntegrationTest.java (hunk 9 above)
Comment (mbarto) [reply src/wms/src/test/java/org/geoserver/wms/wms_1_3/CapabilitiesIntegrationTest.java]: Yes, I am going to add that (and verify the 1.1.1 behaviour related to advertising)

File: src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/CapabilitiesTest.java (hunk 8 above)
Comment (aaime) [reply src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/CapabilitiesTest.java]: That is strange, the advertised flag is processed at the catalog level, not directly in the capabilities... but it may be that the two caps generators are accessing the catalog in different ways, causing different behavior.

Comment (mbarto) [reply src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/CapabilitiesTest.java]: The 1.3.0 transformer has an explicit filter for enabled AND advertised, in the 1.1.1 only I can only see one for enabled, I will check if the advertised flag eventually works at a lower (catalog) level

Comment (aaime) [reply src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/CapabilitiesTest.java]: That should only be perforamnce related, the AdvertisedCatalog class should be filtering away layers that are not advertised regardless (and there might well be a bug in there)

**Comment by mbarto** [comment]
@aaime @jodygarnett most of your feedbacks should be implemented now

File: doc/en/user/source/services/wms/reference.rst [hunk 13]
```diff
@@ -119,6 +119,10 @@ They are fully documented in the :ref:`wms_vendor_parameters` section.
    * - ``format``
      - No
      - request the capabilities document in a certain format
+   * - ``rootLayer``
+     - No
+     - Flag to enable/disable the standard Root top level Layer element. Values are true or false. When false, the Root element will be included only if there is NOT a single top level element already.
```
Comment (aaime) [inline doc/en/user/source/services/wms/reference.rst]:
```suggestion
     - Flag to enable/disable the standard Root top level Layer element. Values are true or false. When false, the Root element will be included only if there are multiple top level layers, if there is only one, it will be the root layer itself.
```

File: doc/en/user/source/services/wms/vendor.rst [hunk 14]
```diff
@@ -326,4 +326,34 @@ An example request:
...
+
+  
+rootLayer
+---------
+
+The ``rootLayer`` parameter can be used to request capabilities documents to include or not a top level root Layer container.
+By default this top level root is always included as a parent of the configured layers and groups. The default can be changed at the service (WMS) level,
+or at the layer / group level.
+
+Using this parameter it is possible to exclude the default root when the resulting document already has a single top element (e.g. a single group with nested children).
```
Comment (aaime) [inline doc/en/user/source/services/wms/vendor.rst]:
```suggestion
Using this parameter it is possible to exclude the default root when the resulting document has a single top element (e.g. a single group with nested children).
```

File: doc/en/user/source/services/wms/webadmin.rst [hunk 15]
```diff
@@ -23,6 +23,11 @@ Root Layer Information
 
 In this section is possible to define a title and an abstract for the root layer in the WMS capabilities. When these are left empty the WMS service title and abstract are used.
 
+It is also possible to set the flag **Always include Root Layer in capabilities**.
+This is checked by default, but can be unset so that the root layer is included in capabilities only when there is NOT already a single top level Layer element.
+This can be useful to allow compatibility with some WMS clients that are not happy with the two or more Layer tree levels.
```
Comment (aaime) [inline doc/en/user/source/services/wms/webadmin.rst]:
```suggestion
This can be useful to allow compatibility with some WMS clients that are not happy with the two or more layer tree levels.
```

**Comment by aaime** [comment]
I've added small doc suggestions, looks otherwise good to me

File: doc/en/user/source/data/webadmin/layergroups.rst [hunk 16]
```diff
@@ -128,4 +128,23 @@ Is possible to associate a layer group with some keywords that will be used to a
   
    Layer groups keywords editor
 
-Layer groups keywords will no be merged with contained layers keywords but keywords of a layer group should be logically inherited by contained layers.
\ No newline at end of file
+Layer groups keywords will no be merged with contained layers keywords but keywords of a layer group should be logically inherited by contained layers.
```
Comment (bradh) [inline doc/en/user/source/data/webadmin/layergroups.rst]: no be -> not be.
